### PR TITLE
Vulkan Memory Allocator

### PR DESCRIPTION
**Pull request by thatcosmonaut**

### CODE REVIEW

**Comment by thatcosmonaut** [comment]
Ready for review!

File: src/FNA3D_Driver_Vulkan.c [hunk 1]
```diff
@@ -2635,154 +2627,199 @@ static void VULKAN_INTERNAL_CreateBackingBuffer(
...
+	VkDeviceSize totalPhysicalSize, totalAllocated, alignment;
+	uint32_t i;
+	VulkanSubBuffer *subBuffer;
 
-		renderer->vkUnmapMemory(
-			renderer->logicalDevice,
-			buffer->deviceMemory
-		);
+	/* Which physical buffer should we suballocate from? */
+	for (i = 0; i < PHYSICAL_BUFFER_MAX_COUNT; i++)
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:2723]: `++`->`+= 1`

File: src/FNA3D_Driver_Vulkan.c [hunk 2]
```diff
@@ -2795,135 +2832,85 @@ static void VULKAN_INTERNAL_SetBufferData(
...
-	}
+	return (FNA3D_Buffer*) result;
+}
 
-	result->usageFlags = usageFlags;
+static void VULKAN_INTERNAL_MaybeExpandStagingBuffer(
+	VulkanRenderer *renderer,
+	VkDeviceSize size
+) {
+	if (size <= renderer->textureStagingBuffer->size) { return; }
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:2905]: Can be expanded to multiline

File: src/FNA3D_Driver_Vulkan.c [hunk 3]
```diff
@@ -6818,34 +6744,36 @@ static void VULKAN_ApplyVertexBufferBindings(
 	renderer->numVertexBindings = numBindings;
 	renderer->currentVertexBufferBindingHash = hash;
 
-	firstVertexBufferIndex = MAX_BOUND_VERTEX_BUFFERS * renderer->currentFrame;
 	bufferCount = 0;
-	for (i = 0; i < numBindings; i += 1)
-	{
-		vertexBufferIndex = firstVertexBufferIndex + i;
 
+	for (i = 0; i < numBindings; i++)
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6749]: `++`->`+= 1`

File: src/FNA3D_Driver_Vulkan.c [hunk 4]
```diff
@@ -8154,26 +7913,24 @@ static void VULKAN_GetVertexBufferData(
...
+		subbuf->physicalBuffer->mapPointer + subbuf->offset + offsetInBytes,
 		elementCount * vertexStride
 	);
 
 	if (useStagingBuffer)
 	{
 		src = cpy;
 		dst = dataBytes;
-		for (i = 0; i < elementCount; i += 1)
+		for (i = 0; i < elementCount; i++)
```
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:7933]: `++`->`+= 1`

**Pull request merged**